### PR TITLE
Exponential correlation function with variable power

### PR DESCRIPTION
**Pull request by yqliaohk**
I added an option to use exponential correlation function with a variable power.
I kept the original `squar_exp` and `abs_exp` to maintain potential backward compatibility, but I think these three can be combined and re-organized to reduce duplication if this won't introduce backward incompatibility or it is not a concern.
I can work on combining them if the smt team agrees this is ideal.
The tests passed locally. Exceptions are using this exponential function with a power other than 2 for `test_kpls_auto.py`. But I also noticed it failed when using original `abs_exp`. If this is expected, I can leave as it is. If not, I can try to fix it maybe with someone on team to help me better understand this test.
Please let me know any suggestions on additional tests.

### CODE REVIEW

**Comment by relf** [comment]
Hi Yingqian, thanks for your contribution. Indeed it is a generalization and it would be nice to refactor and remove code duplication. 

Nevertheless we have to assess the impact of changing the string enumeration (if we go down this line we could also use a Python enum). I am always a bit reluctant to loose backward compatibility, but at the same time we are preparing a 2.0, so it is definitely the right time to break things! 😅

**Review by relf — CHANGES_REQUESTED** [review]
Ok, I suggest to keep backward compatibility while using your generalized implementation and remove duplication.

* First I would rename the new value `exp` in `pow_exp` (as there is also an `act_exp`) and the option `power` in `pow_exp_power` to be more specific.
* On initialization, if `abs_exp` or `squar_exp` is set, then automatically set `corr` option to `pow_exp` and option `pow_exp_power` with relevant power value.
* You noticed that some conditionals are done against `squar_exp` or `abs_exp` meaning they have to be replaced by the equivalent tests on `pow_exp_power`.

**Comment by yqliaohk** [comment]
Hi Rémi, thank you for your review! These are great suggestions. I will make corresponding changes. (Expected time: by Mar 5)

**Comment by yqliaohk** [comment]
I have tried refactoring them based on the suggestions. I had some difficulties to implement exactly the same for the second suggestion because the option can still be changed after initialization, so I think I still need to keep the value update part after initialization. Let me know what you think and if you have other suggestions!

File: smt/surrogate_models/kplsk.py [hunk 1]
```diff
@@ -23,13 +23,25 @@ def _initialize(self):
...
 
+        declare(
+            "pow_exp_power",
+            2.0,
+            desc="Power for the pow_exp kernel function",
+            types=(float),
+        )
+
+
     def _componentwise_distance(self, dx, opt=0, theta=None, return_derivative=False):
```
Comment (Paul-Saves) [inline smt/surrogate_models/kplsk.py:34]: The value pow_exp_power should be > 0 and =<2.

The kernel function $\phi(x) = \exp [- \theta |x|^p]$ is positive definite for a given positive $\theta$ if $0<p \leq 2$ [Corollary~3]{Schoenberg}. METRIC SPACES AND POSITIVE DEFINITE FUNCTIONS* BY I. J. SCHOENBERG

**Review by Paul-Saves — CHANGES_REQUESTED** [review]

File: smt/surrogate_models/kplsk.py (hunk 1 above)
Comment (yqliaohk) [reply smt/surrogate_models/kplsk.py]: Thank you for pointing it out. I added a check. Please let me know if any additional change is needed.

File: smt/surrogate_models/krg_based.py [hunk 2]
```diff
@@ -78,6 +81,12 @@ def _initialize(self):
             ),
             desc="Correlation function type",
         )
+        declare(
+            "pow_exp_power",
+            3.0,
```
Comment (Paul-Saves) [inline smt/surrogate_models/krg_based.py:86]: Can you initialize to a value other than 3 ? Let's say 1.5 ?

File: smt/surrogate_models/kplsk.py (hunk 1 above)
Comment (Paul-Saves) [reply smt/surrogate_models/kplsk.py]: Thanks a lot !

File: smt/surrogate_models/krg_based.py (hunk 2 above)
Comment (yqliaohk) [reply smt/surrogate_models/krg_based.py]: Yes! sorry I tested different default values locally but forgot to change it back.

**Review by Paul-Saves — APPROVED** [review]

File: smt/surrogate_models/krg_based.py [hunk 3]
```diff
@@ -78,6 +81,12 @@ def _initialize(self):
             ),
             desc="Correlation function type",
         )
+        declare(
+            "pow_exp_power",
+            1.9,
+            types=(float),
+            desc="Power for the pow_exp kernel function, values (0.0, 2.0], for squar, abs, and matern, this will be initialized correspondingly by default",
```
Comment (relf) [inline smt/surrogate_models/krg_based.py:88]:
```suggestion
            desc="Power for the pow_exp kernel function (valid values in ]0.0, 2.0]). This option is set automatically when corr option is squar, abs or matern.",
```

**Review by relf — CHANGES_REQUESTED** [review]

File: smt/applications/mfk.py [hunk 4]
```diff
@@ -130,6 +130,12 @@ def _initialize(self):
         )
         self.name = "MFK"
 
+        # initialize default power values
+        if self.options["corr"] == "squar_exp":
```
[the remark this answers is not in the document]
Comment (yqliaohk) [reply smt/applications/mfk.py:134]: Good catch! sorry I forgot about this one.

File: smt/surrogate_models/kpls.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def _initialize(self):
         declare(
             "corr",
             "squar_exp",
-            values=("abs_exp", "squar_exp"),
+            values=("abs_exp", "squar_exp"), # Should add pow_exp?
```
[the remark this answers is not in the document]
Comment (yqliaohk) [reply smt/surrogate_models/kpls.py:28]: I add tests in `test_kpls.py`, let me know if you think that isn't enough.

**Review by relf — APPROVED** [review]

**Comment by relf** [comment]
@yqliaohk Thanks for your work and this nice contribution!

**Comment by yqliaohk** [comment]
> @yqliaohk Thanks for your work and this nice contribution!

My pleasure to contribute! Thank you both for the comments and suggestions!

**Pull request merged**